src/java/org/apache/cassandra/index/sai/plan/Expression.java

```diff
@@ -256,37 +258,44 @@ public boolean isSatisfiedBy(ByteBuffer columnValue)
 
     private boolean validateStringValue(ByteBuffer columnValue, ByteBuffer requestedValue)
     {
+        AbstractAnalyzer analyzer = analyzerFactory.get();
```

Comment:
Is there a reason we're getting a fresh analyzer every time here? Why can't we just leave the previous `analyzer` field and initialize it from the factory at object construction?

Reply:
`AbstractAnalyzer` has the term state.

Reply:
Why wasn't it a problem then previously, when the analyzer instance was reused?

Reply:
I don't see any code that reuses the analyzer.

Reply:
I meant, earlier it was a field, so each call reused the same instance.

Reply:
Show me the code you're referring to.

Reply:
See changes on line 116:
```
-        this.analyzer = columnContext.getAnalyzer();
+        this.analyzerFactory = columnContext.getQueryAnalyzerFactory();
```

Earlier, the `analyzer` was a field initialized at construction. You replaced it with a factory, and now the analyzer is constructed later, on every call to `validateStringValue`. Why can't the analyzer be constructed in the constructor, as it was before? Has something fundamentally changed in the analyzer that now it has state, and earlier it didn't?

Why can't line 116 simply become:
```
this.analyzer = columnContext.getQueryAnalyzerFactory().get();
```
?

Reply:
`AbstractAnalyzer` has a stateful `next` variable, see: https://github.com/datastax/cassandra/blob/ds-trunk/src/java/org/apache/cassandra/index/sai/analyzer/AbstractAnalyzer.java#L45

Therefore `AbstractAnalyzer` cannot be opened once and reused.

Reply:
Previously the code worked because a new analyzer was created each time, however it also reparsed all the options each time as well.

```    
public AbstractAnalyzer getAnalyzer()
{
        Map<String, String> options = config != null ? config.options : Collections.emptyMap();
        return AbstractAnalyzer.fromOptions(getValidator(), options);
}
```

Reply:
But previously the `getAnalyzer()` was called in the constructor, so it was reused, but reset before use.
Previous constructor:

```java
    public Expression(ColumnContext columnContext)
    {
        this.context = columnContext;
        this.analyzer = columnContext.getAnalyzer();
        this.validator = columnContext.getValidator();
    }
```

Previous `validateStringValue()`:
```java
  private boolean validateStringValue(ByteBuffer columnValue, ByteBuffer requestedValue)
    {
        analyzer.reset(columnValue.duplicate());
        ...
```

Stateful `next` didn't pose any problems because the analyzer was reset.
Why can't we reset it the same way as before?

Reply:
I saw you changed it. Resolving.